apps/encryption/lib/keymanager.php

```
@@ -0,0 +1,511 @@
+
+		$encryptedKey = $this->crypt->symmetricEncryptFileContent($keyPair['privateKey'],
+			$password);
+
+		if ($encryptedKey) {
+			$this->setPrivateKey($uid, $encryptedKey);
+			$this->config->setAppValue('encryption', 'recoveryAdminEnabled', 0);
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * @param $userId
+	 * @param $key
+	 * @return bool
+	 */
+	public function setPublicKey($userId, $key) {
+		return $this->keyStorage->setUserKey($userId, $this->publicKeyId, $key);
+	}
+
+	/**
+	 * @param $userId
+	 * @param $key
+	 * @return bool
+	 */
+	public function setPrivateKey($userId, $key) {
+		return $this->keyStorage->setUserKey($userId,
+			$this->privateKeyId,
+			$key);
+	}
+
+	/**
+	 * write file key to key storage
+	 *
+	 * @param string $path
+	 * @param string $key
+	 * @return boolean
+	 */
+	public function setFileKey($path, $key) {
+		return $this->keyStorage->setFileKey($path, $this->fileKeyId, $key);
+	}
+
+	/**
+	 * set all file keys (the file key and the corresponding share keys)
+	 *
+	 * @param string $path
+	 * @param array $keys
+	 */
+	public function setAllFileKeys($path, $keys) {
+		$this->setFileKey($path, $keys['data']);
+		foreach ($keys['keys'] as $uid => $keyFile) {
+			$this->setShareKey($path, $uid, $keyFile);
+		}
+	}
+
+	/**
+	 * write share key to the key storage
+	 *
+	 * @param string $path
+	 * @param string $uid
+	 * @param string $key
+	 * @return boolean
+	 */
+	public function setShareKey($path, $uid, $key) {
+		$keyId = $uid . '.' . $this->shareKeyId;
+		return $this->keyStorage->setFileKey($path, $keyId, $key);
+	}
+
+	/**
+	 * Decrypt private key and store it
+	 *
+	 * @param string $uid userid
+	 * @param string $passPhrase users password
+	 * @return ISession
+	 */
+	public function init($uid, $passPhrase) {
+		try {
+			$privateKey = $this->getPrivateKey($uid);
+			$privateKey = $this->crypt->decryptPrivateKey($privateKey,
+				$passPhrase);
+		} catch (PrivateKeyMissingException $e) {
```

@schiesbn I'd prefer to not catch these exceptions - we need to fail hard in case anything is not properly setup in encryption - data is easily mangled. THX